tests/test_cold_demand.py

```diff
@@ -64,11 +67,102 @@ def test_airco(self):
+        2. energy conservation in the network
+        3. heat to discharge (note cold line is colder than T_ground)
+        4. the cyclic heat_stored contraint, which ensures yearly heat balance between the warm and
+        cold well
+
+        """
+        import models.wko.src.example as example
+        from models.wko.src.example import HeatProblem
+
+        base_folder = Path(example.__file__).resolve().parent.parent
+
+        class HeatingCoolingProblem(HeatProblem):
+            def constraints(self, ensemble_member):
+                constraints = super().constraints(ensemble_member)
+
+                # Add cyclic constraint, this will also ensure that the total power used in the
+                # lower temp & higher temp well and via versa is the same.
+                # Note:
+                #   - WKO in cooling mode: Hot well is being charged with heat and the cold well is
+                # being discharged
+                #   - WKO in heating mode: Cold well is being charged and the hot well is being
+                #     discharged.
+                for a in self.energy_system_components.get("low_temperature_ates", []):
+                    stored_heat = self.state_vector(f"{a}.Stored_heat")
+                    constraints.append(((stored_heat[-1] - stored_heat[0]), 0.0, 0.0))
+                    # This was added to force the heatpump to start loading the WKO 1st 3 timesteps
+                    constraints.append((stored_heat[0], 0.0, 0.0))
+
+                # TODO: confirm if the yearly balance between warm and cold well is heat / flow
+                # related?
+                # This code below might be needed
+                # Add constriant such that the total volume used of the lower temp & higher temp
+                # well of the WKO is the same. Yearly volume balance between the 2 wells. This
+                # ensures volume balance over the timeline. This is done by adding a constriant to
+                # esnure the total volume change over the period = 0.
+                # Note:
+                #   - Volume increase: Hot well is being charged and the cold well is being
+                #     discharged. -> WKO in cooling mode
+                #   - Volume decrease: Cold well is being charged and the hot well is being
+                #     discharged. -> WKO in heating mode
+                # if heat balance over the year is required, then this needs to change ?????? TBC
+                # for ates_id in self.energy_system_components.get("low_temperature_ates", []):
+                #     stored_volume = self.state_vector(f"{ates_id}.Stored_volume")
+                #     volume_usage = 0.0
+                #     for itstep in range(len(self.times()) - 1):
+                #         volume_usage = (
+                #             volume_usage + stored_volume[itstep + 1] - stored_volume[itstep]
+                #         )
+                #     constraints.append((volume_usage, 0.0, 0.0))
```

Comment:
instead of looping over the timesteps, you could consider to set the value of the volume at step 0 and the last step equal. Then it uses the end volume as an initial state at the start of the year.

Reply:
I agree

Reply:
Code update. Currently it is commented out because one has to choose between either heat balance or volume balance over the time period